gomods.go

```diff
@@ -0,0 +1,227 @@
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(info)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "mod":
+		mod, err := dp.GoMod(r.Context(), m.Name, m.Version)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(mod)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "zip":
+		zip, err := dp.Zip(r.Context(), m.Name, m.Version)
+		if err != nil {
+			return err
+		}
+		defer zip.Close()
+		w.Write([]byte{})
+		_, err = io.Copy(w, zip)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "latest":
+		info, err := dp.Latest(r.Context(), m.Name)
+		if err != nil {
+			return err
+		}
+		json, err := json.Marshal(info)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(json)
+		if err != nil {
+			return err
+		}
+		return nil
+	default:
+		return fmt.Errorf("the requested file's extension is not supported")
+	}
+}
+
+func (m Module) fetch(r *http.Request, c Config) (download.Protocol, error) {
+	fs := afero.NewOsFs()
+	fetcher, err := module.NewGoGetFetcher(c.Gomods.GoBinary, fs)
+	if err != nil {
+		return nil, err
+	}
+	s, err := m.storage(c)
+	if err != nil {
+		return nil, err
+	}
+	dp := m.dp(fetcher, s, fs, c)
+	return dp, nil
+}
+
+func (m Module) storage(c Config) (storage.Backend, error) {
+	switch c.Gomods.Cache.Type {
+	case "local":
+		s, err := fs.NewStorage(c.Gomods.Cache.Path, afero.NewOsFs())
+		if err != nil {
+			return nil, fmt.Errorf("could not create new storage from os fs (%s)", err)
+		}
+		return s, nil
+	}
+	return nil, fmt.Errorf("Invalid storage config for gomods")
+}
+
+func (m Module) dp(fetcher module.Fetcher, s storage.Backend, fs afero.Fs, c Config) download.Protocol {
+	lister := download.NewVCSLister(c.Gomods.GoBinary, fs)
+	st := stash.New(fetcher, s, stash.WithPool(c.Gomods.Workers), stash.WithSingleflight)
+	dpOpts := &download.Opts{
+		Storage: s,
+		Stasher: st,
+		Lister:  lister,
+	}
+	dp := download.New(dpOpts, addons.WithPool(c.Gomods.Workers))
+	return dp
+}
+
+func moduleNameAndVersion(path string) (string, string, string) {
+	pathSlice := strings.Split(path, "/")[1:] // [1:] ignores the empty slice item
+	var ext, version string
+	for k, v := range pathSlice {
+		if v == "@v" {
+			ext = gomodsRegex.FindAllStringSubmatch(pathSlice[k+1], -1)[0][0]
+			break
+		}
+	}
+	moduleName := strings.Join(pathSlice[0:3], "/")
+	if strings.Contains(path, "@latest") {
+		return strings.Join(pathSlice[0:3], "/"), "", "latest"
+	}
+	if !strings.Contains(path, "list") {
+		version = modVersionRegex.FindAllStringSubmatch(path, -1)[0][1]
+	}
+	return moduleName, version, ext
+}
+
+// ParseGomods parses the txtdirect config for gomods
+func (gomods Gomods) ParseGomods(c *caddy.Controller, key, value string) error {
+	switch value {
+	case "enable":
+		value, err := strconv.ParseBool(value)
+		if err != nil {
+			return c.ArgErr()
+		}
+		gomods.Enable = value
+	case "gobinary":
+		if value == "" {
+			value = os.Getenv("GOROOT") + "/bin/go"
```

Comment:
Setting a default value is good, but this would only be the default, when `gobinary` is added to the config. So a general default seems better.